### PR TITLE
Fixing slow pipeline tests

### PR DESCRIPTION
**Pull request by Narsil**
Some tests were broken because of pytorch `inference_mode`.
This should cover all cases of `inplace` tensor modifications afaik.

Let me know if there are better ways to fix those.

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@stas00
@patrickvonplaten

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/unispeech/modeling_unispeech.py [hunk 1]
```diff
@@ -947,7 +947,10 @@ def _conv_out_length(input_length, kernel_size, stride):
         return input_lengths
 
     def _get_feature_vector_attention_mask(self, feature_vector_length: int, attention_mask: torch.LongTensor):
-        output_lengths = self._get_feat_extract_output_lengths(attention_mask.sum(-1)).to(torch.long)
+        # Effectively attention_mask.sum(-1), but not inplace to be able to run
+        # on inference mode.
+        mask = attention_mask.cumsum(dim=-1)[:, -1]
```
Comment (patrickvonplaten) [inline src/transformers/models/unispeech/modeling_unispeech.py:952]: (nit) - could we call it `non_padded_lengths` - the idea here is to extract the sub sampled length from the "real non-padded" input length
```suggestion
        non_padded_lengths = attention_mask.cumsum(dim=-1)[:, -1]
```

File: src/transformers/models/unispeech_sat/modeling_unispeech_sat.py [hunk 2]
```diff
@@ -948,7 +948,10 @@ def _conv_out_length(input_length, kernel_size, stride):
         return input_lengths
 
     def _get_feature_vector_attention_mask(self, feature_vector_length: int, attention_mask: torch.LongTensor):
-        output_lengths = self._get_feat_extract_output_lengths(attention_mask.sum(-1)).to(torch.long)
+        # Effectively attention_mask.sum(-1), but not inplace to be able to run
+        # on inference mode.
+        mask = attention_mask.cumsum(dim=-1)[:, -1]
```
Comment (patrickvonplaten) [inline src/transformers/models/unispeech_sat/modeling_unispeech_sat.py:953]: same as above

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 3]
```diff
@@ -989,7 +989,10 @@ def _conv_out_length(input_length, kernel_size, stride):
         return input_lengths
 
     def _get_feature_vector_attention_mask(self, feature_vector_length: int, attention_mask: torch.LongTensor):
-        output_lengths = self._get_feat_extract_output_lengths(attention_mask.sum(-1)).to(torch.long)
+        # Effectively attention_mask.sum(-1), but not inplace to be able to run
+        # on inference mode.
+        mask = attention_mask.cumsum(dim=-1)[:, -1]
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:994]: same as above

File: src/transformers/pipelines/table_question_answering.py [hunk 4]
```diff
@@ -93,76 +93,74 @@ def __init__(self, args_parser=TableQuestionAnsweringArgumentHandler(), *args, *
...
-        with torch.no_grad():
-            return self.model(**inputs)
+        return self.model(**inputs)
 
     def sequential_inference(self, **inputs):
         """
         Inference used for models that need to process sequences in a sequential fashion, like the SQA models which
         handle conversational query related to a table.
         """
-        with torch.no_grad():
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/table_question_answering.py:104]: Nice!

**Review by patrickvonplaten — APPROVED** [review]
Would be happy if we could give `mask` a better naming. Apart from that, thanks a lot for enabling inference mode for all models :-)

File: src/transformers/models/unispeech/modeling_unispeech.py (hunk 1 above)
Comment (Narsil) [reply src/transformers/models/unispeech/modeling_unispeech.py:952]: True, do you know any other way to do that operation ? it's very surprising that `.sum` is inplace, and I am scared that using `cumsum` instead is super wasteful.

Tried to grep it in our code, but I couldn't find anything of that sort.

File: src/transformers/models/detr/modeling_detr.py [hunk 5]
```diff
@@ -648,9 +648,10 @@ def forward(
         hidden_states = residual + hidden_states
         hidden_states = self.final_layer_norm(hidden_states)
 
-        if torch.isinf(hidden_states).any() or torch.isnan(hidden_states).any():
-            clamp_value = torch.finfo(hidden_states.dtype).max - 1000
-            hidden_states = torch.clamp(hidden_states, min=-clamp_value, max=clamp_value)
+        if self.training:
+            if torch.isinf(hidden_states).any() or torch.isnan(hidden_states).any():
+                clamp_value = torch.finfo(hidden_states.dtype).max - 1000
+                hidden_states = torch.clamp(hidden_states, min=-clamp_value, max=clamp_value)
```
Comment (Narsil) [inline src/transformers/models/detr/modeling_detr.py:654]: @stas00 IS that ok to remove at inference time ?

Comment (stas00) [reply src/transformers/models/detr/modeling_detr.py:654]: In theory yes. In practice, it depends on how the model was pre-trained.

The model weights don't change during inference, so we don't need to keep things in check all the time.

However if the pre-trained model's weights lead to an overflow in a single iteration during training, as is the case with some mt5 models under mixed-precision then this can occur just as well during inference.

This is primarily an issue with pre-trained on bf16 models fine-tuned/inferenced on fp16 (mixed or non-mixed precision).

If a model was pretrained with fp16/mixed precision it's pretty sure the clamping won't be needed.

To give you a more intelligent answer it'd require running some tests with the actual DETR models and checking its activations magnitudes, which should be pretty trivial, using https://huggingface.co/transformers/debugging.html#underflow-and-overflow-detection, which can be plugged into HF Trainer and the examples.

File: src/transformers/models/unispeech/modeling_unispeech.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/models/unispeech/modeling_unispeech.py]: Hmm - don't really know to be honest...`torch.sum(...)` doesn't work either? But I think using `.cumsum(...)` is totally fine as well

File: src/transformers/models/detr/modeling_detr.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/transformers/models/detr/modeling_detr.py:654]: To be honest I think this code was just badly copy pasted, so I'm more in favor of disabling this hack for training (as it is done now)

File: src/transformers/models/unispeech/modeling_unispeech.py (hunk 1 above)
Comment (Narsil) [reply src/transformers/models/unispeech/modeling_unispeech.py]: No, `torch.sum(..)` doesn't work.

File: src/transformers/models/detr/modeling_detr.py (hunk 5 above)
Comment (Narsil) [reply src/transformers/models/detr/modeling_detr.py:654]: Ok, if everyone is favorable, then let's do this.

**Comment by patrickvonplaten** [comment]
Good for merge for me

File: src/transformers/models/detr/modeling_detr.py (hunk 5 above)
Comment (stas00) [reply src/transformers/models/detr/modeling_detr.py:654]: > To be honest I think this code was just badly copy pasted, so I'm more in favor of disabling this hack for training (as it is done now)

you must have meant for inference, right Patrick?
